### ISSUE TITLE
Morris - PickUp Scheduling Rules

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2426]
_Imported from trac ticket 2426.
Created by **dwarren**
Opened in trac: Jul 26, 2013, 1:51:24 PM
Last modified in trac: Aug 19, 2013, 4:48:32 PM_
***
_by dwarren:_
Morris will use a universal zone (PUP) to identify pickups filling same day from the written store (accessories, etc).  In store pickups requiring transfer will use the store's pickup zone.  Morris will utilize the "sync pickup capacity" functionality introduced in #2419 to insure that transfer capacity is not exceeded.  To support this the following rules are needed:
 - When PUP is selected, 
   - Default the written store into the "from site" Field. ([#2428](2428))  
   - Only allow "From Site" changes within the following groupings: (00, 11), (30, 35, 12), (32, 36, 13).  Also allow store 01 to assign any "From Site" when choosing the PUP zone.
   - Allow saving if the order can't fully fill.
   - Don't allow saving if any lines fill from another site outside the groupings.
 - Move PUP to be the first selection
 - No modifications needed when selecting the store's pickup zone

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Jul 29, 2013, 10:14:36 AM:_

Mark Bracey wrote:
We need to also add 01 as a written store allowed to reserve from any store if they choose PUP as the zone. 
(Added this to subpoint 2 in original Description, sj)

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Jul 30, 2013, 9:49:17 AM:_

Needs to work with [#2428](2428)

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 30, 2013, 10:38:18 AM:_

Note: _Only allowing save if any lines that are filled are filled out of set of sites_ makes the following requirement redundant in my mind: _Only allow "From Site" changes_.  I plan to not code any From Site change restrictions unless I can understand need, so if I'm missing something, please let me know.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 30, 2013, 11:43:29 AM:_

(Changeset [[4389]](../commit/7f07f1814a8cf0f3b6d1eb995126279c77661f6e))
* closes #2426, Morris PickUp Scheduling Rules surrounding PUP zone #mor

**Comment by kentbower** [comment]
_**kbower** trac status change on Jul 30, 2013, 11:43:29 AM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 1, 2013, 11:30:49 AM:_

(Changeset [[4397]](../commit/43e6a5be951be8a941477bdb9d9f9922c270821e))
* Sort PUP zone at the top of the zones list for Morris Pickups (re #2426) #mor

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Aug 1, 2013, 11:49:38 AM:_

11, 12, and 13 are outlet and were removed from the store groupings [[4391]](../commit/a41d9678d5ff97ce38b5eee218a454f8a6784378)

Please don't forget to add this one:
Mark Bracey wrote:
We need to also add 01 as a written store allowed to reserve from any store if they choose PUP as the zone.

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 1, 2013, 2:39:58 PM:_

That's the one where we asked if the shuttle pickups to 01 and they said rarely, and said they should put it on separate order. We can make that change if they want it, but I could see it getting abused. 

Kent

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Aug 1, 2013, 4:29:54 PM:_

Not exactly, I think that means they want to allow anyone logged into store 01 to save a PU using the PUP zone from any store.  I think not allowing a mix of other stores and 01 on a PUP still stands.

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 2, 2013, 10:16:25 AM:_

It seems irrelevant which store you are logged into (it should be irrelevant anyway).  PUP means "local pickup."  
The only time there would be a need for this is if it was going to need to be shuttled to 01.

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 2, 2013, 11:28:46 AM:_

(Changeset [[4401]](../commit/44a65aacb4926a88fe383928a8d4c98252559db9))
* fix order of the rule execution re #2426, so checks are made after filling #mor

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 6, 2013, 1:14:01 PM:_

1. When I zones (transfer zones) are used for pickups, rule needed to prevent locally filled inventory (unless transfer has already taken place, as indicated by closed pickupcapacity).
1. When PUP is used, do not use priority fill.  Rather, only fill from local "From Site"

**Comment by kentbower** [comment]
_**kbower** trac status change on Aug 6, 2013, 1:14:01 PM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 6, 2013, 1:15:08 PM:_

(Changeset [[4409]](../commit/edec59e07a3e2c7fabe2f199e4c7d697913ac864))
* closes #2426 which was reopened for these: #mor
   * When I zones (transfer zones) are used for pickups, rule needed to prevent locally filled inventory (unless transfer has already taken place, as indicated by closed pickupcapacity). 
   * When PUP is used, do not use priority fill. Rather, only fill from local "From Site"

**Comment by kentbower** [comment]
_**kbower** trac status change on Aug 6, 2013, 1:15:08 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 9, 2013, 3:00:02 PM:_

SQL to help Morris
```
Pickups that currently use these transfer zones but on dates where there is will be no transfer need to be opened and changed to a date there will actually be a transfer.  This SQL shows you which orders are using pickup points for a date that doesn't exist in DEL:

-- PICKUPS THAT NEED DELIVERY DATE CHANGED
select to_char(deliverydate, 'DD-MON-YYYY') as DELIVERY, shipzoneid, orderid
from arc.orders o
where deliverytype = 'P'
and status = 'O'
and shipzoneid in ('I00','I07','I20','I21','I22','I30','I33','I36','I40','I41','I42')
and deliverydate between trunc(sysdate) and to_date('01-JAN-2020','DD-MON-YYYY')
and not exists (select 0 from arc.deliverycapacities dc where dc.zoneid = o.shipzoneid and dc.deliverydate = o.deliverydate)
order by deliverydate, shipzoneid, orderid


This report finds lines that are scheduled on these transfer zones, but are filled locally (already at the fromsite) and therefore should be part of 'PUP' zone, not an I zone.

-- PICKUP LINES CONSUMING POINTS ON OPEN TRANSFER ZONES WITH LOCALLY FILLED INVENTORY
select o.orderid, l.lineid, o.shipzoneid, to_char(o.deliverydate, 'DD-MON-YYYY') as DELIVERY
from arc.orders o, arc.orderdetails l
where o.deliverytype = 'P'
and o.status = 'O'
and o.shipzoneid in ('I00','I07','I20','I21','I22','I30','I33','I36','I40','I41','I42')
and o.orderid = l.orderid
and l.fillsiteid is not null
and not exists (select 0 from arc.deliverycapacities dc where dc.zoneid = o.shipzoneid and dc.deliverydate = o.deliverydate and dc.closed = 'Y')
and o.deliverydate >= trunc(sysdate)
and decode(l.fillsiteid, '35', '30', '36', '32', l.fillsiteid) = decode(o.fromsiteid, '35', '30', '36', '32', o.fromsiteid)
order by o.orderid, o.shipzoneid, l.lineid


PUP points records that have no corresponding DEL date and are not currently in use and therefore safe to be deleted.  Theoretically these should be filled = 0

-- PUP CAPACITIES THAT ARE SAFE TO BE DELETED (RE-CHECK IMMEDIATELY BEFORE DELETING)
select to_char(deliverydate, 'DD-MON-YYYY') as "DATE", zoneid, filled
from arc.pickupcapacities pc
where zoneid in ('I00','I07','I20','I21','I22','I30','I33','I36','I40','I41','I42')
and not exists (select 0 from arc.deliverycapacities dc where dc.zoneid = pc.zoneid and dc.deliverydate = pc.deliverydate)
and deliverydate between trunc(sysdate) and to_date('01-JAN-2020','DD-MON-YYYY')
and not exists (select 0 from arc.orders o
    where o.deliverytype = 'P'
    and o.status = 'O'
    and pc.zoneid = o.shipzoneid
    and pc.deliverydate = o.deliverydate)
order by deliverydate, zoneid



PUP points records that have no corresponding DEL date and so need to be deleted before go live (and after making sure they are not in use)!


-- PUP CAPACITIES THAT NEED TO BE DELETED
select to_char(deliverydate, 'DD-MON-YYYY') as "DATE", zoneid, filled
from arc.pickupcapacities pc
where zoneid in ('I00','I07','I20','I21','I22','I30','I33','I36','I40','I41','I42')
and not exists (select 0 from arc.deliverycapacities dc where dc.zoneid = pc.zoneid and dc.deliverydate = pc.deliverydate)
and deliverydate between trunc(sysdate) and to_date('01-JAN-2020','DD-MON-YYYY')
order by deliverydate, zoneid

-- PUP CAPACITIES THAT NEED TO BE ADDED
select to_char(deliverydate, 'DD-MON-YYYY') as "DATE", zoneid, filled
from arc.deliverycapacities d
where zoneid in ('I00','I07','I20','I21','I22','I30','I33','I36','I40','I41','I42')
and not exists (select 0 from arc.pickupcapacities p where p.zoneid = d.zoneid and p.deliverydate = d.deliverydate)
and deliverydate between trunc(sysdate) and to_date('01-JAN-2020','DD-MON-YYYY')
order by deliverydate, zoneid
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 9, 2013, 3:03:38 PM:_

Changeover notes:
* mor dataupdate: 
  1. SQL to guarantee no PUP missed 
  1. sum PUP cap to corresponding DEL
  1. enable sync
  1. then set PUP to DEL (use trigger and update statement)

**Comment by kentbower** [comment]
_**kbower** trac status change on Aug 9, 2013, 3:03:51 PM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 19, 2013, 4:48:32 PM:_

(Changeset [[4450]](../commit/44cf752bae11821c48cefec99bd96471d4fe92ed))
* closes #2426, SQL update statements for capacity conversion to implement mirroring pickup capacities #mor

**Comment by kentbower** [comment]
_**kbower** trac status change on Aug 19, 2013, 4:48:32 PM:_ **closed**
